### PR TITLE
Add Jetson Nano Support

### PR DESCRIPTION
**Pull request by dwalkes**
See thread starting at https://hub.mender.io/t/nvidia-tegra-jetson-tx2/123/10

Remaining steps:
 - [x] Run the integration checklist on Nano
 - [x] Resolve issue described [here](https://hub.mender.io/t/nvidia-tegra-jetson-tx2/123/18)
 - [x] Test on TX2

### CODE REVIEW

**Comment by Ecordonnier** [comment]
Just a detail, the commit message should be changed "Replace CONFIG_ENV_IS_NOWHERE with a CONFIG_ENV_IS_NOWHERE configuration"

It is obviously a typo and is supposed to be "Replace CONFIG_ENV_IS_NOWHERE with a CONFIG_ENV_IS_MMC configuration"

**Comment by dwalkes** [comment]
@moqmar thanks for these changes!

I tried to test on TX 2 this weekend but ran into 2 problems which I don't think are related to your changes.

This first problem: https://github.com/Trellis-Logic/meta-mender-community/pull/8 keeps me from being able to run tegraflash, attempts to run hang when writing the APP partition.  The workaround linked in that pull request gets around the issue for me, I still don't yet understand what changed or why this is necessary.

I have a second problem which results in this message on the serial terminal after `mender-install-manual` and redundant bootloader update:
```
[0000.147] C> Error: mts ratchet check failed
```
The boot halts after this, before even starting u-boot.
I'm not sure where this is coming from.  I rolled back to the last meta-tegra commit I thought I previously tested successfully, https://github.com/madisongh/meta-tegra/commit/6d41181d397984a9261558ab019827c51777286b and I see the same thing, with or without Nano changes included.  

A `mender -install` on the command line without redundant boot component update works fine.

So, in summary, I think I'm convinced the problems I'm seeing on TX2 aren't related to your Nano changes.

@Ecordonnier I see your [comment on mender hub](https://hub.mender.io/t/nvidia-tegra-jetson-tx2/123/20) stating you tested mender update on tegra.  Can you confirm this was without `mender-install-manual` and redundant boot support?  Have you tried a tegraflash on TX2 yet?  I expect you'll need the change in  https://github.com/Trellis-Logic/meta-mender-community/pull/8 for this.

**Comment by Ecordonnier** [comment]
@dwalkes  I tested the update on tx2 WITH redundant boot support / mender-install-manual, however I tested by flashing the tegraflash image (script doflash.sh), and I was testing with a custom image based on jetson-tx2. Are you testing with a custom image or with a standard yocto image?

**Comment by dwalkes** [comment]
@Ecordonnier 

Thanks for the response.

> Are you testing with a custom image or with a standard yocto image?

I was using core-image-minimal, based on a standard yocto image and the [instructions in the meta-mender-community repository and repo tool.](https://github.com/mendersoftware/meta-mender-community/tree/sumo/meta-mender-tegra).  I've previously used on a custom build successfully.

I just realized one difference from the last time I tried on the yocto image is I probably used core-image-base the first time I tried with Yocto.  I'll give that a try to see if I notice any difference.

**Comment by Ecordonnier** [comment]
FYI I am trying to make the branch work, as you know what still need to be done is to write the dataimg file to the data partition correctly. I pushed current state of my attempt to do this at https://github.com/taurob/meta-mender-community  (branch feature/jetson-nano). I am able to boot jetson-nano, but I can't mount the dataimg partition, so something is wrong:

> bash-4.4# mount /dev/mmcblk0p15 /tmp 
> [ 1629.535764] EXT4-fs (mmcblk0p15): bad geometry: block count 507904 exceeds size of device (32768 blocks)
> mount: /tmp: wrong fs type, bad option, bad superblock on /dev/mmcblk0p15, missing codepage or helper program, or other error.

I installed gptdisk manually to jetson-nano, and this is my partition table (looks good as far as I can tell):

> Command (? for help): p
> Disk /dev/mmcblk0: 62333952 sectors, 29.7 GiB
> Sector size (logical/physical): 512/512 bytes
> Disk identifier (GUID): 4A6B9FCA-0837-474A-BCF8-9C0F1E6F1186
> Partition table holds up to 128 entries
> Main partition table begins at sector 2 and ends at sector 33
> First usable sector is 34, last usable sector is 33554398
> Partitions will be aligned on 2048-sector boundaries
> Total free space is 19518 sectors (9.5 MiB)
> 
> Number  Start (sector)    End (sector)  Size       Code  Name
>    1         4483072        33554398   13.9 GiB    8300  APP
>    2            2048            2303   128.0 KiB   8300  TBC
>    3            4096            4991   448.0 KiB   8300  RP1
>    4            6144            7295   576.0 KiB   8300  EBT
>    5            8192            8319   64.0 KiB    8300  WB0
>    6           10240           10623   192.0 KiB   8300  BPF
>    7           12288           13439   576.0 KiB   8300  TOS
>    8           14336           14463   64.0 KiB    8300  EKS
>    9           16384           17663   640.0 KiB   8300  LNX
>   10           18432           19327   448.0 KiB   8300  DTB
>   11           20480           20735   128.0 KiB   8300  RP4
>   12           22528           22687   80.0 KiB    8300  BMP
>   13           24576           24959   192.0 KiB   8300  ENV
>   14           26624         4220927   2.0 GiB     8300  APP_B
>   15         4220928         4483071   128.0 MiB   8300  DATA
> 


There are occurences of "507904" in meta-mender, so I will try to investigate tomorrow what this error message means exactly.

**Comment by dwalkes** [comment]
> I just realized one difference from the last time I tried on the yocto image is I probably used core-image-base the first time I tried with Yocto. I'll give that a try to see if I notice any difference.

core-image-base works fine, so this is just a problem with core-image-minimal.  I've added a note to the [mender hub page known issues section](https://hub.mender.io/t/nvidia-tegra-jetson-tx2/123).

**Comment by dwalkes** [comment]
Thanks @Ecordonnier

> I pushed current state of my attempt to do this at https://github.com/taurob/meta-mender-community (branch feature/jetson-nano)

Thanks for sharing!

> I am able to boot jetson-nano, but I can't mount the dataimg partition, so something is wrong:

The partitioning looks different than how it works on TX2 and I haven't had a chance to get into that difference in detail yet.  Do you understand how/where the sizes are getting filled in for each partition?  The TX2 uses size variables in flash_l4t_t186.custom.xml like [APPSIZE](https://github.com/mendersoftware/meta-mender-community/blob/sumo/meta-mender-tegra/recipes-bsp/tegra-binaries/files/flash_l4t_t186.custom.xml#L151) for this, which is [setup in a flash.xml.in file based on the meta-tegra image types class](https://github.com/madisongh/meta-tegra/blob/ce1e2d37793ac07ae20a8b436b828535d768802e/classes/image_types_tegra.bbclass#L187).  The data partition on TX2 uses a [special 0x808 allocation attribute](https://github.com/mendersoftware/meta-mender-community/blob/sumo/meta-mender-tegra/recipes-bsp/tegra-binaries/files/flash_l4t_t186.custom.xml#L423) to [assign all remaining space on the device to the data partition](https://docs.nvidia.com/jetson/l4t/index.html#page/Tegra%2520Linux%2520Driver%2520Package%2520Development%2520Guide%2Fpart_config.html%23wwpID0EMHA).

It looks like the [sdcard-layout.mender.in](https://github.com/Trellis-Logic/meta-mender-community/blob/1bb00950553a5d9798a12e52b08ee295608f0d72/meta-mender-tegra/recipes-bsp/u-boot/sdcard-layout-mender.in) has hard coded values for these and I'm wondering if this might be causing a disconnect about where the data partition is located or how it's sized.  For instance based on this section:

```
14,APP_B,2147483648,APPFILE
15,DATA,134217728,DATAFILE
1,APP,0,APPFILE
```

I think this means APP_B is fixed at 2G, DATA is fixed at 128MB and APP takes up the remaining space in the flash.  This appears to match your gptdisk output

```
Number Start (sector) End (sector) Size Code Name
1 4483072 33554398 13.9 GiB 8300 APP
...
14 26624 4220927 2.0 GiB 8300 APP_B
15 4220928 4483071 128.0 MiB 8300 DATA
```

I'm guessing this error:
```
[ 1629.535764] EXT4-fs (mmcblk0p15): bad geometry: block count 507904 exceeds size of device (32768 blocks)
```
means the DATA partition is larger than the amount of blocks you have reserved for it in the partition?  What does `ls` return for the size of your .dataimg file?  In my core-image base build it's `134217728` 

```
dan@dan-ubuntu:~/mender-tegra$ ls -lsa build/tmp/deploy/images/jetson-tx2/*.dataimg*
348 -rw-r--r-- 1 dan dan 134217728 Dec 10 07:35 build/tmp/deploy/images/jetson-tx2/core-image-base-jetson-tx2--20191210143431.dataimg
```

which I don't think would fit in this space if sector size is 2K (127,268,864 bytes):

```
Number Start (sector) End (sector) Size Code Name
15 4220928 4483071 128.0 MiB 8300 DATA
```

**Comment by Ecordonnier** [comment]
Your understanding of the partitions is correct. 0 is a special value which means "maximum possible size". The .dataimg file is also 134217728 bytes big for my image, which is exactly 128 MiB, and exactly 32768 blocks of 4096 bytes. I didn't have time to work on this today, but the error shown (block count 507904 exceeds size of device (32768 blocks)) means that something is taking 2 080 374 784 bytes (507904*4096) instead of 128MiB. I am not expert in those low-level things, I checked the ext4 filesystem of the .dataimg file and it seems to have the correct size.
I will try to do a clean build, maybe something in my build files still has the old value of 2147483648 for the data partition (2147483648 is the value used in the original branch).

**Comment by Ecordonnier** [comment]
The problem is best shown with the dumpe2fs command. This is on my PC:
```
$ dumpe2fs ogrip-robot-jetson-nano.dataimg
...
Inode count:              32768
Block count:              131072
...
Block size:               1024
```

And this is on target:

```
# dumpe2fs -h /dev/mmcblk0p15
...
Inode count:              507904
Block count:              507904
...
Block size:               4096
```

I'm trying to fix it.


Edit: I was right in my previous post, I deleted tmp and made a new build and now the partition is mounting fine, so it seems that it is only a build dependency issue.

Edit 2: I started working on the next issue:
```
root@jetson-nano:~# fw_printenv 
Cannot access MTD device /dev/mmcblk0boot1: No such file or directory
```

**Comment by Ecordonnier** [comment]
The remaining issue is that on jetson-tx2, the u-boot environment is stored in the boot partition of the emmc (/dev/mmcblk0boot1), but on jetson-nano which is using a sd-card, there is no emmc boot partition and therefore the configuration needs to be changed.
I started making the changes to move u-boot environment to its own partition at the beginning of the sd-card, and updated my branch on github, but it is not done completely (the branch does not compile, u-boot-tegra fails to compile with "#error CONFIG_ENV_OFFSET is not the same as MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_1".

**Comment by moqmar** [comment]
Hi guys! Unfortunately I've been sick the last couple of days, but I'm trying to get this to work right now again. :)

Fixed the commit message & data partition, but there's still the issue left that partition sizes are currently hardcoded.

**Comment by Ecordonnier** [comment]
Hi Moqmar, I am not sure what you mean. I am working on moving the u-boot environment to the sd-card. As far as I understand other than that the partitions are working fine?

**Comment by moqmar** [comment]
Basically, I'm trying to improve the workaround in https://github.com/taurob/meta-mender-community/commit/5967daa5c0dddcc32e0e5b8acbde6345332b629d so one can use a different partition size than 2GB (which is currently hardcoded in sdcard-layout-mender.in).

I haven't tried your latest commit yet, but moving around the partitions at the start always yielded an unbootable sd card for me, hence the out-of-order layout. I think ideally, the APP partitions should have an adjustable size (using IMAGE_ROOTFS_SIZE?!), and DATA should be at the end so it can be resized using the mender-growfs-data feature.

And sorry if the force push with the corrected commit message broke your fork, I think you'll have to do a rebase there...

**Comment by Ecordonnier** [comment]
OK, no problem about the rebase, that's very easy to do. The reason I moved the u-boot environment to the beginning, is that the "environment offset" variable of u-boot is 32 bits, so it cannot be more than a few GB after the start of the sd-card. I sent you an email (to your github email) with my phone number, if you want to discuss this quickly.

**Comment by moqmar** [comment]
Yes, I read about that too - I think it works well if it's just after the system partitions though (which are all quite small), just before the APP/DATA partitions.

**Comment by Ecordonnier** [comment]
hm, ok, that could work. It will just make the offset calculation a bit more complicated.
About your change to the partition size: you need to make sure that the GPT secondary table is at the offset where the primary gpt partition table expects it. I'm not sure it works if you make the last partition fixed-size instead of "0" to fill up the rest of the sd-card.

**Comment by moqmar** [comment]
I'm currently trying the approach from https://github.com/madisongh/meta-tegra/issues/188#issuecomment-545308006, will let you know if that worked out. The system partitions are hardcoded anyways, so that shouldn't be an issue.

**Comment by moqmar** [comment]
Alright, now I'm a bit stuck at the following error message:
```
Failed to read the current active partition: No match between boot and root
partitions.: Failed mender_saveenv_canary check. There is an error in the U-Boot
setup. Likely causes are: 1) Mismatch between the U-Boot boot loader environment
location and the location specified in /etc/fw_env.config. 2) 'mender_setup' is
not run by the U-Boot boot script: exit status 1
```

I manually set `echo '/dev/mmcblk0p13 0 0x2000' > /etc/fw_env.config`, as suggested in the approach linked above - will have to add that to a recipe too.
I'm guessing that `mender_setup` is not being run? Or how can I check the environment directly in U-Boot?

**Comment by Ecordonnier** [comment]
I'm not sure what the state of your branch exactly is. By default, u-boot environment is configured to be in the mmcblk0boot1 partition, which does not exist on jetson-nano. Changing /etc/fw_env.config will only change the config in the rootfs, but u-boot will still be looking at the wrong place. That's why I started to configure "MENDER_UBOOT_CONFIG_SYS_MMC_ENV_PART_tegra210 = "0"", etc. in my tmp commits.

**Comment by dwalkes** [comment]
> Or how can I check the environment directly in U-Boot?

I haven't tried on Nano yet but on jetson-tx2 you can use the u-boot console over a serial port cable.  Probably similar to [these instructions](https://www.jetsonhacks.com/2019/04/19/jetson-nano-serial-console/).

Then from the u-boot console you can use commands printenv and setenv as described in the [integration checklist](https://docs.mender.io/2.2/devices/yocto-project/bootloader-support/u-boot/integration-checklist) or use u-boot commands (for the TX2 `mtdparts` is useful) to verify associated partitions.

Regarding the u-boot location changes it might be helpful to review the recent pull request for warrior which moved them on TX2 from the image into mmcblk0boot1. 
 See https://github.com/mendersoftware/meta-mender-community/pull/107.  Some of these changes might need to be backed out to support a different partition layout in Nano.  Let me know if you have any specific changes there.

I don't have a Jetson Nano currently but I might be able to get my hands on one next week.

**Comment by moqmar** [comment]
Alright, managed to fix the error message with @Ecordonnier's approach. I'm currently trying it out and it seems like most things are working... It seems like I'm missing `nvbootctrl` though, which comes from [tegra186-redundant-boot](https://github.com/madisongh/meta-tegra/search?q=nvbootctrl&unscoped_q=nvbootctrl):

```
2019-12-12 14:04:25 +0000 UTC error: stderr collected while running script /var/lib/mender/scripts/ArtifactReboot_Enter_50 [Starting nvidia mender enter reboot redundant boot steps
/var/lib/mender/scripts/ArtifactReboot_Enter_50: line 5: nvbootctrl: command not found
]
2019-12-12 14:04:25 +0000 UTC debug: status reported, response 204 No Content
2019-12-12 14:04:25 +0000 UTC error: transient error: error calling enter script for (error) reboot state: error running enter state script(s) for ArtifactReboot_Enter state: statescript: error executing 'ArtifactReboot_Enter_50': 127 : exit status 127
```

Other than that, it seems to work quite well already. The rootfs is being updated (which is a bit weird, shouldn't it roll back on failure?), and mender says that it is running from the correct partition, but the kernel seems to still be the old one. The [Directory Module](https://hub.mender.io/t/directory/325) is working perfectly, too.

**Comment by Ecordonnier** [comment]
I can confirm that the update is working (I'm attaching some logs showing a stand-alone update from git version "0c61fac-dirty" to "2ffe707-dirty" (of the repo importing meta-mender-community as submodule) which shows a boot on /dev/mmcblk0p1, and then a boot on /dev/mmcblk0p15 after the reboot. I don't see an error related to nvbootctrl when doing a stand-alone update with the command "mender -install", are you testing using a mender server? I think on jetson-nano we should not call nvbootctrl at all, as long as the update of the bootloader is not implemented (redundant boot).

Also, I find your comment regarding the fact that the kernel is not updated surprising, because the kernel is located in the /boot directory of the rootfs and is not a separated partition on jetson-nano. Therefore if the rootfs gets updated the kernel should be updated too.

[log-update-mender.txt](https://github.com/Trellis-Logic/meta-mender-community/files/3959894/log-update-mender.txt)

**Comment by moqmar** [comment]
You're right, seems like the kernel indeed did update, I just did something wrong when disabling a module. \o/

So, it seems like the best thing to do would be to just disable the `redundant-boot-reboot-state-script` on the Jetson Nano, I'm currently trying to figure out how to do that. Other than that, it seems like everything's working. :smile:

@dwalkes You can check off the second task, I also generally checked the things listed in the integration checklist except for rollback (they all seem to be working fine), but couldn't do it directly as shown in the instructions as I don't have a serial cable here. Also, I don't have a TX2 to test with.

**Comment by Ecordonnier** [comment]
How are you testing? I don't see the redundant-boot-reboot-state script being called in my logs.

**Comment by moqmar** [comment]
This is when using the Mender Server - the file doesn't exist under the specified path though, so I'm a bit unsure where that error is coming from... The final deployment status is `Failure: start executing script: ArtifactReboot_Enter_50`

**Comment by Ecordonnier** [comment]
I wonder if maybe your build state is not correct, because in commit "Remove BUP Payload dependency" you indeed removed the dependency to "mender-tegra-bup-payload-install" which is deploying ArtifactReboot_Enter_50, but maybe the file was still left in your tmp folder.

What do you have there?
```
root@jetson-nano:/etc/mender# ls scripts/
version
root@jetson-nano:/etc/mender# ls /data/mender/scripts/
version

```

**Comment by moqmar** [comment]
Oh, you're right, `bitbake -c clean tegra-state-scripts` fixed the issue, thanks a lot! 😃

**Comment by Ecordonnier** [comment]
OK great! I also checked the values in /etc/fw_env.config and it seems to be consistent with the ENV partition start sector reported by gdisk:

```
root@jetson-nano:~# cat /etc/fw_env.config 
/dev/mmcblk0 0xc00000 0x20000
/dev/mmcblk0 0xc20000 0x20000

```

gdisk shows:
`  13           24576           24959   192.0 KiB   8300  ENV`

And 24576*512 is 12582912 which is 0xc00000, so I think the u-boot environment is not overwriting anything it should not.

I'm going to test on jetson-tx2 today (for testing I am using commit https://github.com/taurob/meta-mender-community/commit/b91ba8a1a85732d9000b6bf32506664e2b979d22 which is your branch rebased on top of origin/warrior, where I just changed the commit message of my commit "tmp2"). Will you send a pull-request to the upstream repo?

**Comment by Ecordonnier** [comment]
I confirm that stand-alone updates are working fine with commit b91ba8a1a85 of meta-mender-community and 0c4783ebbf5606 of meta-mender (warrior branch) on jetson-tx2. I tested using mender-install-manual, which also updated the bootloader.

**Comment by dwalkes** [comment]
Thank you @Ecordonnier and @moqmar!

I see the latest commits in the branch at https://github.com/taurob/meta-mender-community/commit/b91ba8a1a85732d9000b6bf32506664e2b979d22 and changes at https://github.com/Trellis-Logic/meta-mender-community/pull/9

I've added a couple of review comments there, I'm curious about the u-boot environment offset.

I also noticed [this note](https://github.com/Trellis-Logic/meta-mender-community/pull/7#issuecomment-565411841)

> I also generally checked the things listed in the integration checklist except for rollback (they all seem to be working fine), but couldn't do it directly as shown in the instructions as I don't have a serial cable here. 

I think it will be important to do the u-boot steps and test rollback to be sure it's working.  My concern (see code review comments above) is that you might not be writing the ENV partition and u-boot environment might be overwriting other partitions used for tegra boot.

I think the next step would be either for @Ecordonnier to open a PR for nano suport on meta-mender-community using the taurob branch or @moqmar to merge the taorb branch into his repo and open the PR from there.

Note that you'll need to follow the [Contributing](https://github.com/mendersoftware/mender/blob/master/CONTRIBUTING.md) guidelines (see signing and commit format instructions there).

Can either or both of you support the Nano going forward?  If so, please add your names to the Maintainer section of [the readme](https://github.com/mendersoftware/meta-mender-community/tree/warrior/meta-mender-tegra)

Finally, I think we'll need to either add a new mender hub page for the Nano, linked from https://hub.mender.io/t/nvidia-tegra-jetson-tx2/123, or merge the Nano into that page.  Perhaps we should ask Mirza what they'd prefer.

I'm planning to understand and fix https://github.com/Trellis-Logic/meta-mender-community/pull/8 next week, and also try the Nano changes myself.

**Comment by Ecordonnier** [comment]
@moqmar  today I cleaned-up the branch and reduced it to a single commit with both our "sign-off" (https://github.com/Trellis-Logic/meta-mender-community/commit/76b99ed7a2b8fd11d935c835521d1503b7fceda1). I also created the documentation at https://hub.mender.io/t/nvidia-tegra-jetson-nano/1360

I cannot commit to being a maintainer of the jetson-nano more than a few weeks for various reasons, so I did not add my name to the maintainer's list. Is it OK for you if I send a pull request with this commit? You can also do it if you prefer.

**Comment by moqmar** [comment]
Sure, go ahead! 👍 
I won't be able to officially maintain it either though, as I'm only working with the Jetson until March for research purposes, but it's great that it's working so far right now. 😃

**Comment by Ecordonnier** [comment]
@moqmar Great, I created https://github.com/mendersoftware/meta-mender-community/pull/119 and you are the author of the commit.

By the way one question: do you need the rootfs to be 4GB? Or is it just an arbitrary number?

**Comment by dwalkes** [comment]
Closed in favor of https://github.com/mendersoftware/meta-mender-community/pull/119
